### PR TITLE
DnsNameResolver.resolve*(...) never notifies the Future when empty ho…

### PR DESCRIPTION
**Pull request by normanmaurer**
…stname is used.

Motivation:

When an empty hostname is used in DnsNameResolver.resolve*(...) it will never notify the future / promise. The root cause is that we not correctly guard against errors of IDN.toASCII(...) which will throw an IllegalArgumentException when it can not parse its input. That said we should also handle an empty hostname the same way as the JDK does and just use "localhost" when this happens.

Modifications:

- If the try to resolve an empty hostname we use localhost
- Correctly guard against errors raised by IDN.toASCII(...) so we will always noify the future / promise
- Add unit test.

Result:

DnsNameResolver.resolve*(...) will always notify the future.

### CODE REVIEW

File: resolver-dns/src/main/java/io/netty/resolver/dns/DnsNameResolverContext.java [hunk 1]
```diff
@@ -510,11 +512,26 @@ private void followCname(InetSocketAddress nameServerAddr, String name, String c
 
         final InetSocketAddress nextAddr = nameServerAddrs.next();
         if (parent.isCnameFollowARecords()) {
-            query(nextAddr, new DefaultDnsQuestion(cname, DnsRecordType.A), promise);
+            if (!query(hostname, DnsRecordType.A, nextAddr, promise)) {
```
Comment (netkins) [inline resolver-dns/src/main/java/io/netty/resolver/dns/DnsNameResolverContext.java]: MAJOR Merge this if statement with the enclosing one. [rule](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1066)

Comment (normanmaurer) [inline resolver-dns/src/main/java/io/netty/resolver/dns/DnsNameResolverContext.java]: I actually think its easier to read the way it is... So I would love to just leave it as it is. @Scottmitch thoughts ?

File: resolver-dns/src/main/java/io/netty/resolver/dns/DnsNameResolver.java [hunk 2]
```diff
@@ -507,6 +512,12 @@ protected void doResolve(String inetHost,
             return;
         }
 
+        if (inetHost.isEmpty()) {
```
Comment (Scottmitch) [inline resolver-dns/src/main/java/io/netty/resolver/dns/DnsNameResolver.java]: the JDK also resolves `null` to localhost too

Comment (Scottmitch) [inline resolver-dns/src/main/java/io/netty/resolver/dns/DnsNameResolver.java]: we should move this check above `createByteArrayFromIpAddressString` which may throw a NPE if `null` is passed in.

File: resolver-dns/src/main/java/io/netty/resolver/dns/DnsNameResolverContext.java (hunk 1 above)
Comment (Scottmitch) [inline resolver-dns/src/main/java/io/netty/resolver/dns/DnsNameResolverContext.java]: I don't think it impacts readability much either way ... less conditionals & indentation if they are combined. minor issue though your call.

File: resolver-dns/src/main/java/io/netty/resolver/dns/DnsNameResolver.java (hunk 2 above)
Comment (normanmaurer) [inline resolver-dns/src/main/java/io/netty/resolver/dns/DnsNameResolver.java]: @Scottmitch yes it does but not our DefaultNameResolver where we disallow null: https://github.com/netty/netty/blob/4.1/resolver/src/main/java/io/netty/resolver/SimpleNameResolver.java#L58

We could remove this restriction tho... WDYT ?

Comment (Scottmitch) [inline resolver-dns/src/main/java/io/netty/resolver/dns/DnsNameResolver.java]: I guess it is inline with this PR which is making behavior more consistent with JDK's resolution.

Comment (normanmaurer) [inline resolver-dns/src/main/java/io/netty/resolver/dns/DnsNameResolver.java]: @Scottmitch maybe we even should just move the empty and null handling to `SimpleNameResolver` ?

Comment (Scottmitch) [inline resolver-dns/src/main/java/io/netty/resolver/dns/DnsNameResolver.java]: @normanmaurer - sgtm. I'm not sure if @trustin would have any opinion on this though.

Comment (normanmaurer) [inline resolver-dns/src/main/java/io/netty/resolver/dns/DnsNameResolver.java]: @trustin WDYT ?

Comment (trustin) [inline resolver-dns/src/main/java/io/netty/resolver/dns/DnsNameResolver.java]: Not to `SimpleNameResolver` but to `InetNameResolver`? `SimpleNameResolver` can be used for non-INET scenarios.

File: resolver-dns/src/main/java/io/netty/resolver/dns/DnsNameResolverContext.java (hunk 1 above)
Comment (trustin) [inline resolver-dns/src/main/java/io/netty/resolver/dns/DnsNameResolverContext.java]: I think it reads better this way and one less indentation level is a bonus.

```java
if (parent.isCnameFollowARecords() &&
    !query(hostname, DnsRecordType.A, nextAddr, promise)) {
    ...
}
```

File: resolver-dns/src/main/java/io/netty/resolver/dns/DnsNameResolver.java (hunk 2 above)
Comment (vietj) [inline resolver-dns/src/main/java/io/netty/resolver/dns/DnsNameResolver.java]: +1 for being consistent with JDK's resolution

**Comment by normanmaurer** [comment]
@Scottmitch PTAL again... This was the only way I could find without introduce some sort of API breaking in non `@UnstableApi` marked stuff.

File: resolver-dns/src/test/java/io/netty/resolver/dns/DnsNameResolverTest.java [hunk 3]
```diff
@@ -496,6 +497,47 @@ public void testResolveIp() {
...
+
+    @Test
+    public void testResolveEmptyIpv6() {
+        testResolveEmpty(InternetProtocolFamily.IPv6, NetUtil.LOCALHOST6);
+    }
+
+    private static void testResolveEmpty(InternetProtocolFamily family, InetAddress expectedAddr) {
+        DnsNameResolver resolver = newResolver(family).build();
+        try {
+            InetAddress address = resolver.resolve("").syncUninterruptibly().getNow();
```
Comment (Scottmitch) [inline resolver-dns/src/test/java/io/netty/resolver/dns/DnsNameResolverTest.java]: also test with `null`?

File: resolver-dns/src/test/java/io/netty/resolver/dns/DnsNameResolverTest.java [hunk 4]
```diff
@@ -496,6 +497,47 @@ public void testResolveIp() {
...
+
+    @Test
+    public void testResolveAllEmptyIpv6() {
+        testResolveAllEmpty(InternetProtocolFamily.IPv6, NetUtil.LOCALHOST6);
+    }
+
+    private static void testResolveAllEmpty(InternetProtocolFamily family, InetAddress expectedAddr) {
+        DnsNameResolver resolver = newResolver(family).build();
+        try {
+            List<InetAddress> addresses = resolver.resolveAll("").syncUninterruptibly().getNow();
```
Comment (Scottmitch) [inline resolver-dns/src/test/java/io/netty/resolver/dns/DnsNameResolverTest.java]: also test with `null`?

File: resolver-dns/src/test/java/io/netty/resolver/dns/DnsNameResolverTest.java (hunk 3 above)
Comment (normanmaurer) [inline resolver-dns/src/test/java/io/netty/resolver/dns/DnsNameResolverTest.java]: yep will do

File: resolver-dns/src/test/java/io/netty/resolver/dns/DnsNameResolverTest.java (hunk 4 above)
Comment (normanmaurer) [inline resolver-dns/src/test/java/io/netty/resolver/dns/DnsNameResolverTest.java]: yep will do

**Comment by normanmaurer** [comment]
Cherry-picked into 4.1 (a416b79d865fd6ce11ecce74d53ce2dfd8be007f)
